src/main.rs

```diff
@@ -782,6 +794,79 @@ fn start_vmm(cmd_arguments: ArgMatches) -> Result<Option<String>, Error> {
     r.map(|_| api_socket_path)
 }
 
+// This is a best-effort solution to the latency induced by the RCU
+// synchronization that happens in the kernel whenever the file
+// descriptor table is full and needs to be expanded. The initial table
+// has 64 entries on amd64.
+// Rust programs that use {File,EventFd}::try_clone() to share handles
+// are impacted by this issue.
+// This behavior is quite noticeable in the restore snapshot scenario,
+// the latency is a big chunk of the total time required to start
+// cloud-hypervisor and to restore the snapshot.
+// The kernel has an optimization in the expand table code path where it
+// doesn't call synchronize_rcu() if there is only one thread in the
+// process.
+// To address this issue we attempt to expand the descriptor table at
+// application start, when it has only one thread.
+fn expand_fdtable() -> Result<(), FdTableError> {
+    let mut limits = libc::rlimit {
+        rlim_cur: 0,
+        rlim_max: 0,
+    };
+
+    // SAFETY: FFI call with valid arguments
+    if unsafe { libc::getrlimit(libc::RLIMIT_NOFILE, &mut limits) } < 0 {
+        return Err(FdTableError::GetRLimit(io::Error::last_os_error()));
+    }
+
+    let mut table_size: i32;
+
+    if limits.rlim_cur < 4096 {
+        table_size = limits.rlim_cur as i32;
+
+        let new_limits = libc::rlimit {
+            rlim_cur: std::cmp::min(4096, limits.rlim_max),
+            rlim_max: limits.rlim_max,
+        };
+
+        // If this fails we are using the current rlim_cur which is lower
+        // than 4096.
+        // SAFETY: FFI call with valid arguments
+        if unsafe { libc::setrlimit(libc::RLIMIT_NOFILE, &new_limits) } == 0 {
+            table_size = new_limits.rlim_cur as i32;
+        }
+    } else {
+        table_size = 4096;
+    }
+
+    if table_size <= 3 {
+        return Ok(());
+    }
+
+    // Test if the file descriptor is empty
+    // SAFETY: FFI call with valid arguments
+    let flags: i32 = unsafe { libc::fcntl(table_size - 1, libc::F_GETFD) };
+    if flags >= 0 {
+        // Nothing to do, the table is already big enough
+        return Ok(());
+    }
+
+    let err = io::Error::last_os_error();
+    if err.raw_os_error() != Some(libc::EBADF) {
+        return Err(FdTableError::GetFd(err));
+    }
+    // SAFETY: FFI call with valid arguments
+    if unsafe { libc::dup2(0, table_size - 1) } < 0 {
```

Comment:
Technically, there might not be a descriptor 0.  (It's not a good idea, but it's not _wrong.)  Maybe it would be better to create a new file descriptor in the most lightweight way possible, and then dup2 it.  (An eventfd, maybe?)

Reply:
Correct, stdin might not always be present, especially in daemon programs. Opening a file won't work as /dev/null or any other file might not even be in your root, that's why I used stdin for that. Eventfd might do the trick, I'll give it a shot.

Reply:
Done, I created an eventfd and used dup2 on that.